### PR TITLE
draft changes to desktop palette

### PR DESCRIPTION
**Pull request by IsaacNoel**
Fixing issue #14

### CODE REVIEW

File: src/gradients-schemes/ScaleScheme.java [hunk 1]
```diff
@@ -67,14 +67,15 @@ else if ( var > min )
 			{
 				double tempval = min - var ;
 				double tempmax = min - max;
-				perc = tempval / tempmax ;
+        tempmax /= size;
```
Comment (brandesNW) [inline src/gradients-schemes/ScaleScheme.java]: make comment explaining that (min - max)/size is the width of each color slice, and that

File: src/gradients-schemes/ScaleScheme.java [hunk 2]
```diff
@@ -67,14 +67,15 @@ else if ( var > min )
...
+				perc = Math.floor(tempval / tempmax);
```
Comment (brandesNW) [inline src/gradients-schemes/ScaleScheme.java]: this calculation determines the correct color slice

File: src/gradients-schemes/ScaleScheme.java [hunk 3]
```diff
@@ -84,11 +85,11 @@ else if ( var < min )
...
 				double tempval = var - min ;
 				double tempmax = max - min ;
-				perc = tempval / tempmax ;
+        tempmax /= (double) size;
+				perc = Math.floor(tempval / tempmax);
 			}
 		}
-		index = (int) Math.ceil((perc * (size - 1))) ;
-
+		index = Math.min((int) perc, size-1);
```
Comment (brandesNW) [inline src/gradients-schemes/ScaleScheme.java]: spaces needed around subtraction operator -

**Review by brandesNW — CHANGES_REQUESTED** [review]
The tabs should be converted to spaces. In Atom with my settings using Edit->Lines->Auto Indent did the trick.
The leading curly braces should be on the line with the statement, not their own line.
I think the logic would be clearer if the normal case (min < max) came first.

Ideally the core logic
```
        tempmax /= (double) size;
	perc = Math.floor(tempval / tempmax);
```
should not be repeated.

for some of the code the ternary operator ?:  could be useful
booleanExpression ? expression1 : expression2

I'm not sure it's worth doing, but the code would be a lot clearer if organized as

// Main case: the value lies between the maximum and minimum
				double tempval = (max > min) ? var - min : min - var;
				double tempmax = abs(max - min );
                                tempmax /= (double) size;
				perc = Math.floor(tempval / tempmax);
                                index = Math.min((int) perc, size-1);

//  Case value maps to first endpoint
  if  (max > min) && ( var < min ) ||  (min > max) && ( var > min) {
     index = 0
  }

//  Case value maps to last endpoint
  if min > max) && ( var < min )  ||  (min > max) && ( var < max) {
     index = size - 1
  }

this error message should be improved

```
					"The number of colors in a scheme is limited to " +
					String.valueOf(ColorSchemes.getRGBArray( schemename , legendname ).length) +
					" but your third argument is " + size) ;
```
In particular it should refer to number of colors ,not third argument.

perhaps the first part should be something like:  legendname + "allows a maximum of " + String.valueOf(ColorSchemes.getRGBArray( schemename , legendname ).length)

**Review by brandesNW — APPROVED** [review]
This needs a separate submission of checksums.txt in order for the model tests to pass.

**Review by LaCuneta — APPROVED** [review]
This looks okay to me.  I ran the language tests in desktop and they passed.  I ran the test `palette:scale-scheme "Divergent" "BrBG" 9 3 0 10` in 6.2.0 with the prior version of Palette and got a different result.  I ran the Heat Diffusion - Alternate Vis model and saw the visual difference.  The exact change is hard for me to judge since I don't know the code very well, but if the existing language tests are passing, I'm good to merge.

I would like to see the text of the pull request itself updated with a description of the problem and the solution it contains, just to have it "on the record".

**Pull request merged**